### ISSUE TITLE
Webpack 4 "--mode production" breaks

### ISSUE THREAD

**Issue by EdmundMai** [developit/preact-compat#484]
For some reason, when I start the server with

`cross-env NODE_ENV=production webpack --config='config/webpack.config.js' --mode production`

it breaks and ReactDOM.render(...) returns `<undefined></undefined>`. Changing it to `--mode development` fixes it.

`cross-env NODE_ENV=production webpack --config='config/webpack.config.js' --mode development`

Webpack config:

```
  module.exports = merge(common, {
    bail: true,
    devtool: shouldUseSourceMap ? "source-map" : false,
    devtool: "cheap-module-source-map",
    entry: paths.appSrc + "/script.js",
    output: {
      path: paths.appBuild,
      filename: `static/js/oyster-webnode-${APP_VERSION}.min.js`,
      chunkFilename: "static/js/[name].chunk.js",
      publicPath: publicPath,
      devtoolModuleFilenameTemplate: info =>
        path
          .relative(paths.appSrc, info.absoluteResourcePath)
          .replace(/\\/g, "/")
    },
    resolve: {
      modules: ["node_modules", paths.appNodeModules].concat(
        process.env.NODE_PATH.split(path.delimiter).filter(Boolean)
      ),
      extensions: [".web.js", ".mjs", ".js", ".json", ".web.jsx", ".jsx"],
      alias: {
        react: "preact-compat",
        "react-dom": "preact-compat"
      },
      plugins: [new ModuleScopePlugin(paths.appSrc, [paths.appPackageJson])]
    },
    plugins: [
      new BundleAnalyzerPlugin({
        generateStatsFile: generateStatsFile
      }),
      new HtmlWebpackPlugin({
        inject: true,
        template: paths.appHtml,
        minify: {
          removeComments: true,
          collapseWhitespace: true,
          removeRedundantAttributes: true,
          useShortDoctype: true,
          removeEmptyAttributes: true,
          removeStyleLinkTypeAttributes: true,
          keepClosingSlash: true,
          minifyJS: true,
          minifyCSS: true,
          minifyURLs: true
        }
      }),
      new InterpolateHtmlPlugin(env.raw)
    ]
  });

```

**Comment by nehaabrol87** [comment]
@developit  This is the same issue I was facing 3 weeks ago.
As I mentioned on the issue there are 5 plugins that mode = production automatically adds
I bet its one of those 5

**Provides process.env.NODE_ENV with value production. Enables FlagDependencyUsagePlugin, FlagIncludedChunksPlugin, ModuleConcatenationPlugin, NoEmitOnErrorsPlugin, OccurrenceOrderPlugin, SideEffectsFlagPlugin and UglifyJsPlugin**

**Comment by tadeuszwojcik** [comment]
Happens for me as well, in production mode, `VNode` lacks properties defined via `Object.defineProperty`,  this is what I got:
```
{
  "children": [...],
  "attributes": {...},
  "preactCompatUpgraded": true,
  "preactCompatNormalized": true
}
```
There is no type of props there. Everything looks fine in dev mode.

**Comment by gerhardsletten** [comment]
I did also had this problem where I was using preact in production. But found that I would need the `reduce_vars` setting for UglifyJSPlugin to making `react-router` work:

```js
const UglifyJsPlugin = require('uglifyjs-webpack-plugin')
// and this config
optimization: {
  minimizer: [
    new UglifyJsPlugin({
      uglifyOptions: {
        compress: {
          reduce_vars: false // see https://github.com/developit/preact/issues/961
        }
      }
    })
  ]
}
```

And with another package `redux-connect` there was a problem because it in the transpiled files was require `babel-runtime/helpers/jsx` instead of `react`

```js
var _jsx2 = require('babel-runtime/helpers/jsx');
```

I am no babel-expert but it looks to me like babel-runtime is a deprecated method to use, so I [upgraded the package](https://github.com/makeomatic/redux-connect/pull/131) and after switching to this my project works fine with preact-compat and webpack 4

**Comment by developit** [comment]
Sorry for the absence. I believe we may have merged a workaround for this Uglify bug in Preact itself, which will fix the issues described here. Sorry for the really difficult debugging!

Here's the PR to Preact:
https://github.com/developit/preact/pull/1153

We'll release it as a patch update shortly.

**Comment by developit** [comment]
This is fixed in [Preact 8.3.0](https://github.com/developit/preact/releases/tag/8.3.0). Update and enjoy!
